test/Validate.CFG.cpp

```diff
@@ -0,0 +1,601 @@
+  Block middle("middle", SpvOpBranchConditional);
+  Block end("end", SpvOpReturn);
+
+  middle.setBody(
+      "%cond    = OpSLessThan %intt %one %two\n"
+      "OpSelectionMerge %end None\n");
+
+  Block entry2("entry2");
+  Block middle2("middle2");
+  Block end2("end2", SpvOpReturn);
+
+  string str = header + nameOps("middle2", make_pair("func", "Main")) + types_consts +
+               "%func    = OpFunction %voidt None %funct\n";
+
+  str += entry >> middle;
+  str += middle >> vector<Block>({end, middle2});
+  str += end;
+  str += "OpFunctionEnd\n";
+
+  str += "%func2    = OpFunction %voidt None %funct\n";
+  str += entry2 >> middle2;
+  str += middle2 >> end2;
+  str += end2;
+  str += "OpFunctionEnd\n";
+
+  CompileSuccessfully(str);
+  ASSERT_EQ(SPV_ERROR_INVALID_CFG, ValidateInstructions());
+  EXPECT_THAT(getDiagnosticString(), HasSubstr("Main"));
+  EXPECT_THAT(getDiagnosticString(), HasSubstr("middle2"));
+  EXPECT_THAT(getDiagnosticString(),
+              HasSubstr("are referenced but not defined"));
+}
+
+TEST_F(ValidateCFG, HeaderDoesntDominatesMergeBad) {
+  Block entry("entry");
+  Block bad("bad", SpvOpBranchConditional);
+  Block f("f");
+  Block merge("merge", SpvOpBranchConditional);
+  Block cont("cont");
+  Block end("end", SpvOpReturn);
+
+  bad.setBody("OpLoopMerge %merge %cont None\n");
+
+  merge.setBody(
+      " %cond    = OpSLessThan %intt %one %two\n"
+      "OpSelectionMerge %end None\n");
+
+  string str = header + nameOps("bad", "merge", make_pair("func", "Main")) +
+               types_consts + "%func    = OpFunction %voidt None %funct\n";
+
+  str += entry >> merge;
+  str += merge >> vector<Block>({bad, end});
+  str += bad >> vector<Block>({cont, f});
+  str += cont >> merge;
+  str += f >> end;
+  str += end;
+  str += "OpFunctionEnd\n";
+
+  CompileSuccessfully(str);
+  ASSERT_EQ(SPV_ERROR_INVALID_CFG, ValidateInstructions());
+  EXPECT_THAT(getDiagnosticString(), HasSubstr("Header block"));
+  EXPECT_THAT(getDiagnosticString(), HasSubstr("doesn't dominate"));
+  EXPECT_THAT(getDiagnosticString(), HasSubstr("[bad]"));
+  EXPECT_THAT(getDiagnosticString(), HasSubstr("[merge]"));
+}
+
+TEST_F(ValidateCFG, UnreachableMerge) {
+  Block entry("entry");
+  Block branch("branch", SpvOpBranchConditional);
+  Block t("t", SpvOpReturn);
+  Block f("f", SpvOpReturn);
+  Block merge("merge");
+  Block end("end", SpvOpReturn);
+
+  branch.setBody(
+      " %cond    = OpSLessThan %intt %one %two\n"
+      "OpSelectionMerge %merge None\n");
+
+  string str = header + nameOps("branch", "merge", make_pair("func", "Main")) +
+               types_consts + "%func    = OpFunction %voidt None %funct\n";
+
+  str += entry >> branch;
+  str += branch >> vector<Block>({t, f});
+  str += t;
+  str += f;
+  str += merge >> end;
+  str += end;
+  str += "OpFunctionEnd\n";
+
+  CompileSuccessfully(str);
+  ASSERT_EQ(SPV_SUCCESS, ValidateInstructions());
+}
+
+TEST_F(ValidateCFG, UnreachableMergeInstruction) {
```

Comment:
The test name is not terribly suggestive of what's being tested, especially when compared to the previous one.